### PR TITLE
core: device pta: add flag to watch for running tee-supplicant

### PR DESCRIPTION
**Pull request by muvarov**
Some TAs require tee-supplicant to be run. For example fTPM requires
storage services provided by tee-supplicant. When scanning and
probe() devices on tee bus we can initialize early drivers which
do not require tee-supplicant and after mount fs and tee-supplicant
run do probe() drivers witch require tee-supplicant.

Signed-off-by: Maxim Uvarov <maxim.uvarov@linaro.org>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
How is this second probe triggered? Do you have any kernel code you can share?

File: core/pta/device.c [hunk 1]
```diff
@@ -71,8 +77,9 @@ static TEE_Result invoke_command(void *pSessionContext __unused,
 				 TEE_Param pParams[TEE_NUM_PARAMS])
 {
 	switch (nCommandID) {
-	case PTA_CMD_GET_DEVICES:
```
Comment (jforissier) [inline core/pta/device.c:74]: This line is still needed

**Comment by muvarov** [comment]
@jenswi-linaro  please find kernel patch in this PR: https://github.com/Linaro/meta-ledge/pull/203

File: core/pta/device.c (hunk 1 above)
Comment (muvarov) [reply core/pta/device.c:74]: thanks, my fault on merge. Fixed.

**Comment by jenswi-linaro** [comment]
The reason fTPM is an early TA is to be able to load it early before tee-supplicant is running, but it depends on tee-supplicant so it can't be accessed until tee-supplicant is running.

This sounds like fTPM shouldn't be an early TA. Instead we need a way of enumerating normal TAs. There's the `ta_ver.db` which holds the uuid and version of all TAs that has been verified once. We could extend it to include a flags field too. I think it might be a bit much to have the kernel probe for all known TAs.

On the other hand, we only expect very few TAs to be matched by a device driver in the kernel so hard coding these uuids in OP-TEE could also be an option. A TA used by a device driver is a contract with the kernel so it's good if OP-TEE OS itself has some control over which TAs are advertised.

**Comment by b49020** [comment]
> The reason fTPM is an early TA is to be able to load it early before tee-supplicant is running, but it depends on tee-supplicant so it can't be accessed until tee-supplicant is running.

If we carefully look at the dependency here, fTPM TA depends on RPMB secure storage service provided by tee-supplicant. If in future we decide to directly route the RPMB secure storage service via Linux kernel then it could very well act as proper early TA. And if we achieve that then it can be utilized to have features like measured boot etc. So IMO, fTPM being an early TA is the best suited place to leverage its complete functionality.

> Instead we need a way of enumerating normal TAs.

Yeah this patch will complement enumeration of normal TAs as well. As we only need to add an extra RPC call to tee-supplicant in `get_devices()` in case invoked via `PTA_CMD_GET_DEVICES_SUPP`. Then tee-supplicant can parse `flags` of normal TAs and return the list of TAs which can be enumerated.

> On the other hand, we only expect very few TAs to be matched by a device driver in the kernel so hard coding these uuids in OP-TEE could also be an option.

AFAIR, there are already 4 known candidate device drivers dependent on this. And we can't predict how many will be there especially considering platform specific Linux kernel services. Also, hard coding is simply not scalable and essentially defeats the purpose of enumeration.

> A TA used by a device driver is a contract with the kernel so it's good if OP-TEE OS itself has some control over which TAs are advertised.

I have a slightly different view here. I think its the Service Provider being the signing authority of TA should decide how a particular TA should behave and OP-TEE OS should be flexible to allow that behaviour.

**Comment by jenswi-linaro** [comment]
>> A TA used by a device driver is a contract with the kernel so it's good if OP-TEE OS itself has some control over which TAs are advertised.

> I have a slightly different view here. I think its the Service Provider being the signing authority of TA should decide how a particular TA should behave and OP-TEE OS should be flexible to allow that behaviour.

So you mean that it's the Service Provider that has the contract with the kernel? For a vendor kernel tree I see how this can work. For an upstream kernel tree it seem more hazardous. Perhaps if all TAs provided by a Service Provider need be presented together with some identifier (uuid?) of the Service Provider.

**Comment by b49020** [comment]
> So you mean that it's the Service Provider that has the contract with the kernel? For a vendor kernel tree I see how this can work. For an upstream kernel tree it seem more hazardous.

Let me try to elaborate my perspective of communication among a Linux device driver and OP-TEE based TA:

- Linux kernel provides a generic TEE bus framework which is expected to be stable and compatible irrespective of underlying TEE implementation like OP-TEE.
- OP-TEE OS provides the functionality to enumerate TAs which can act as devices on TEE bus.
- A TA (pseudo TA or early TA or REE-FS TA) that can act as a TEE device could come either from a silicon vendor (platform specific) or a service provider. If we compare this with other Linux bus eg. PCI bus, devices may come from varying device manufacturers but all complying with base PCI bus enumeration infrastructure.
- There can always be new TEE device drivers for a new set of TA devices (it could either be vendor kernel or upstream kernel) which we aren't aware off as of now. So OP-TEE enumeration should be flexible enough to support those rather than restricting enumeration to hardcoded list of TAs.

> Perhaps if all TAs provided by a Service Provider need be presented together with some identifier (uuid?) of the Service Provider.

Can you please elaborate here? My understanding is that each TA comes with a unique identifier (UUID) which can act as a Linux device.

**Comment by jenswi-linaro** [comment]
There's no central pool from where UUIDs are allocated so two different TAs can by accident (or on purpose) have the same UUID but different ABIs if they are from different Service Providers. This will perhaps not happen in the same system but the Linux kernel may be the same in two different systems with two different SPs.

Right now is optee_os.git in charge of the UUID namespace used for TAs which are exposed as services for device drivers. I'm afraid this is starting to turn into a mess.

**Comment by muvarov** [comment]
> There's no central pool from where UUIDs are allocated so two different TAs can by accident (or on purpose) have the same UUID but different ABIs if they are from different Service Providers. This will perhaps not happen in the same system but the Linux kernel may be the same in two different systems with two different SPs.
> 

UUID is Unique UID isn't it? By design it has to be unique.

> Right now is optee_os.git in charge of the UUID namespace used for TAs which are exposed as services for device drivers. I'm afraid this is starting to turn into a mess.

It might have sense to support tee-bus. Linux kernel or even before linux kernel (uboot, hypervisor and etc). To do this I think we need to have something like current PR does. ftpm is good example which shows that it's better to have it inside optee os then in user TA.

BTW, I'm holding kernel patch for submission until we resolve  this PR

**Comment by jforissier** [comment]
In fact @jenswi-linaro I don't get your argument that "since fTPM depends on `tee-supplicant` it should not be an early TA", why? Of course "supplicant not available" implies "can't be a REE-FS TA, and should therefore probably be an early TA" ; but the opposite is not true?

So I think two things are being discussed here and I'm not sure they are closely related:
1. During enumeration, adding a flag to differentiate {pseudo,early} TAs that need some support from `tee-supplicant` from those that do not;
2. Being able to enumerate REE FS TAs.

I think we should leave (2) for a later discussion and focus on what this PR is trying to achieve, i.e., solve a dependency issue. Unless I missed something.

**Comment by jenswi-linaro** [comment]
> In fact @jenswi-linaro I don't get your argument that "since fTPM depends on `tee-supplicant` it should not be an early TA", why? Of course "supplicant not available" implies "can't be a REE-FS TA, and should therefore probably be an early TA" ; but the opposite is not true?

If fTPM needs tee-supplicant, why not load it with tee-supplicant also? It seems an unnecessary complication to have it as an early TA and then add logic to only load it once tee-supplicant is available.

> 
> So I think two things are being discussed here and I'm not sure they are closely related:
> 
> 1. During enumeration, adding a flag to differentiate {pseudo,early} TAs that need some support from `tee-supplicant` from those that do not;
> 2. Being able to enumerate REE FS TAs.
> 
> I think we should leave (2) for a later discussion and focus on what this PR is trying to achieve, i.e., solve a dependency issue. Unless I missed something.

The discussion has spun off a bit, but is still relevant. We must sort out who's responsible for the ABI towards the kernel driver. If an upstream kernel driver depends on the ABI of a certain UUID we need a way to make that UUID well known so it's not reused by accident (or some other reason) for another ABI.

**Comment by jbech-linaro** [comment]
> > In fact @jenswi-linaro I don't get your argument that "since fTPM depends on `tee-supplicant` it should not be an early TA", why? Of course "supplicant not available" implies "can't be a REE-FS TA, and should therefore probably be an early TA" ; but the opposite is not true?
> 
> If fTPM needs tee-supplicant, why not load it with tee-supplicant also? It seems an unnecessary complication to have it as an early TA and then add logic to only load it once tee-supplicant is available.

As I mentioned [here](https://lkml.org/lkml/2019/7/3/519) I believe it's possible to delay some operations and therefore avoid running it as an early TA. I haven't digged into the technicalities, but it sounds like it should be doable to delay it.

**Comment by muvarov** [comment]
> As I mentioned [here](https://lkml.org/lkml/2019/7/3/519) I believe it's possible to delay some operations and therefore avoid running it as an early TA. I haven't digged into the technicalities, but it sounds like it should be doable to delay it.

In kernel patch connected to this PR (which I pointed to connected PR in ledge) I solved this problem. I.e. ftpm , tee, optee can be compiled in kernel, as modules or combinations. I did not send kernel patch upstream because I think we need to solve optee-os part first.

As I understand from current discussion we have only one open question - why we need early ftpm.ta in this case instead of put it to user ta. I think the answer is to have access to this ta in firmware like uboot. I.e. uboot can use tpm to mount crypted block device and load and start kernel. Then the same ta can be used in linux.

**Comment by b49020** [comment]
> As I mentioned [here](https://lkml.org/lkml/2019/7/3/519) I believe it's possible to delay some operations and therefore avoid running it as an early TA. I haven't digged into the technicalities, but it sounds like it should be doable to delay it.

If you meant to register fTPM device without secure storage operations working, then that implementation is still broken. Consider following scenarios:
- A kernel driver which is dependent on a TPM device may invoke operations like PCR measurement, trusted key seal/unseal etc. These operations may still fail as fTPM NV storage is implemented via RPMB secure storage.
- Similar is scenario with user-space application, if it observes `/dev/tpmX` device active then it could invoke operations and those operations may still fail as tee-supplicant is not active by that time.

So, in summary we should only register fTPM device when secure storage operations are up and running which is something that is achieved via enumerating fTPM TA after tee-supplicant init as per @muvarov implementation.

**Comment by b49020** [comment]
> There's no central pool from where UUIDs are allocated so two different TAs can by accident (or on purpose) have the same UUID but different ABIs if they are from different Service Providers. This will perhaps not happen in the same system but the Linux kernel may be the same in two different systems with two different SPs.

First of all it **isn't** correct to say that two different TA implementations even with same ABI (invoke commands and corresponding params) to have same UUID. And if UUID is same then I would say implementation is **broken** which is true irrespective of whether its a user-space client or a kernel-space client. The UUID is by its nature should be unique for every TA implementation. And for this same reason every Linux driver registers a UUID config table instead of single UUID as per following example:

```
static const struct tee_client_device_id optee_rng_id_table[] = {
        {UUID_INIT(0xab7a617c, 0xb8e7, 0x4d8f,
                   0x83, 0x01, 0xd0, 0x9b, 0x61, 0x03, 0x6b, 0x64)},
        {}
};
```
If we say that this UUID belongs to a specific SP and there is another SP with a different TA implementation, but with same ABI, could simply just extend this config table to list corresponding TA UUID. Also, it is possible to add some quirks to the kernel driver based on which TA UUID is being probed.

So its not only that a particular kernel driver is bound to a single TA but can be compatible with multiple TAs as well given that all those TAs comply to corresponding kernel driver ABI. It is quite similar to the concept of DT compatible property where a single kernel driver can be compatible with multiple DT based devices having different compatible property.

**Comment by b49020** [comment]
> > I think we should leave (2) for a later discussion and focus on what this PR is trying to achieve, i.e., solve a dependency issue. Unless I missed something.
> 
> The discussion has spun off a bit, but is still relevant. We must sort out who's responsible for the ABI towards the kernel driver. If an upstream kernel driver depends on the ABI of a certain UUID we need a way to make that UUID well known so it's not reused by accident (or some other reason) for another ABI.

I agree with @jforissier here. We are somewhat hijacking this discussion to discuss possible evolution of TEE bus with new use-cases coming up. In case my kernel ABI replies are sufficient here, @jenswi-linaro is it fine with you to open a separate issue to discuss it further?

BTW, regarding fTPM TA, as @muvarov mentioned firmware use-cases make it an appropriate candidate to be an early TA.

**Comment by jenswi-linaro** [comment]
I agree, with that last piece the use-case it makes sense to have fTPM as an early TA.

Let's continue the discussion of kernel driver dependencies in a separate issue. @b49020, would you like to start it?

**Comment by b49020** [comment]
> Let's continue the discussion of kernel driver dependencies in a separate issue. @b49020, would you like to start it?

https://github.com/OP-TEE/optee_os/issues/3863

File: lib/libutee/include/user_ta_header.h [hunk 2]
```diff
@@ -23,9 +23,13 @@
 	 * (pseudo-TAs only).
 	 */
 #define TA_FLAG_CONCURRENT		(1 << 8)
-#define TA_FLAG_DEVICE_ENUM		(1 << 9) /* device enumeration */
+	/*
+	 * Device enumeration.
```
Comment (jenswi-linaro) [inline lib/libutee/include/user_ta_header.h:27]: The need some more explanation. For instance:
_Device enumeration is done in two stages by normal world, first before the tee-supplicant has started and then once more when the tee-supplicant is started. The flags below controls if the TA should be reported in the first, second or in both cases._

File: lib/libutee/include/pta_device.h [hunk 3]
```diff
@@ -24,5 +24,6 @@
  * TEE_ERROR_SHORT_BUFFER - Output buffer size less than required
  */
 #define PTA_CMD_GET_DEVICES		0x0
+#define PTA_CMD_GET_DEVICES_SUPP	0x1
```
Comment (jenswi-linaro) [inline lib/libutee/include/pta_device.h:27]: This needs more documentation.

File: core/pta/device.c [hunk 4]
```diff
@@ -21,18 +21,23 @@
 #define PTA_NAME "device.pta"
 
 static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf,
-		   uint32_t blen, uint32_t *pos)
+		   uint32_t blen, uint32_t *pos, bool supplicant)
```
Comment (jenswi-linaro) [inline core/pta/device.c:24]: How about supply the required flags to this function directly instead? Should result in easier logic.

Comment (muvarov) [reply core/pta/device.c:24]: @jenswi-linaro do you mean get_devices(nParamTypes, pParams, nCommandID) then then pass this nCommandID to add_ta() ?

Comment (jenswi-linaro) [reply core/pta/device.c:24]: No, rather
``` C
case PTA_CMD_GET_DEVICES:
        return get_devices(nParamTypes, pParams, TA_FLAG_DEVICE_ENUM);
case PTA_CMD_GET_DEVICES_SUPP:
        return get_devices(nParamTypes, pParams, TA_FLAG_DEVICE_ENUM_SUPP);

Comment (muvarov) [reply core/pta/device.c:24]: yea, that also looks nice.

File: core/pta/device.c [hunk 5]
```diff
@@ -72,7 +78,9 @@ static TEE_Result invoke_command(void *pSessionContext __unused,
 {
 	switch (nCommandID) {
 	case PTA_CMD_GET_DEVICES:
-		return get_devices(nParamTypes, pParams);
+		return get_devices(nParamTypes, pParams, PTA_CMD_GET_DEVICES);
+	case PTA_CMD_GET_DEVICES_SUPP:
+		return get_devices(nParamTypes, pParams, PTA_CMD_GET_DEVICES_SUPP);
```
Comment (jforissier) [inline core/pta/device.c:83]: You have changed here but not `get_devices()` nor `add_ta()`. I doubt it will work well :wink:

File: core/pta/device.c [hunk 6]
```diff
@@ -21,18 +21,20 @@
 #define PTA_NAME "device.pta"
 
 static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf,
-		   uint32_t blen, uint32_t *pos)
+		   uint32_t blen, uint32_t *pos, uint32_t rflags)
 {
-	if (flags & TA_FLAG_DEVICE_ENUM) {
-		if (*pos + sizeof(*uuid) <= blen)
-			tee_uuid_to_octets(buf + *pos, uuid);
+	if (!(flags & rflags & TA_FLAG_DEVICE_ENUM_MASK))
```
Comment (jenswi-linaro) [inline core/pta/device.c:26]: `TA_FLAG_DEVICE_ENUM_MASK` doesn't seem very useful to me.

File: core/pta/device.c [hunk 7]
```diff
@@ -21,18 +21,20 @@
...
+		return;
```
Comment (b49020) [inline core/pta/device.c:27]: I think we need an additional check here to not report a TA which sets both flags `TA_FLAG_DEVICE_ENUM` and `TA_FLAG_DEVICE_ENUM_SUPP`. As it could lead to double registration of same TA as device which seems to be incorrect. And perhaps we could put a warning message to show that its an incorrect combination.

File: core/pta/device.c (hunk 6 above)
Comment (muvarov) [reply core/pta/device.c:26]: I think rflags can have more additional  flags in future and it's better to write here that we are looking only for 2 bits.

File: core/pta/device.c (hunk 7 above)
Comment (muvarov) [reply core/pta/device.c:27]: I think if both flags are specified then TA /driver is written in that way that double initialization is expected. However it might be not easy to think up such use case.

File: core/pta/device.c (hunk 6 above)
Comment (jenswi-linaro) [reply core/pta/device.c:26]: Current there's two callers, one with `TA_FLAG_DEVICE_ENUM` and another with `TA_FLAG_DEVICE_ENUM_SUPP`. What will you do if you add another bit? Probably expanding the mask in order to allow that other bit. To me it seems pretty useless, at least for now.

File: core/pta/device.c [hunk 8]
```diff
@@ -21,18 +21,20 @@
 #define PTA_NAME "device.pta"
 
 static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf,
-		   uint32_t blen, uint32_t *pos)
+		   uint32_t blen, uint32_t *pos, uint32_t rflags)
 {
-	if (flags & TA_FLAG_DEVICE_ENUM) {
-		if (*pos + sizeof(*uuid) <= blen)
-			tee_uuid_to_octets(buf + *pos, uuid);
+	if (!(flags & rflags))
```
Comment (jenswi-linaro) [inline core/pta/device.c:26]: Please use positive logic also, like in the original code.
Avoiding to change lines that doesn't need to be changed is usually a good idea.

File: core/pta/device.c (hunk 7 above)
Comment (b49020) [reply core/pta/device.c:27]: It isn't like the same device will be initialized twice but it will be like two separate devices registered for single TA and each device probed at once which doesn't sounds correct.

While at it, haven't you faced following error while enabling supplicant enumeration:

```
[    2.330622] sysfs: cannot create duplicate filename '/devices/optee-clnt0'
[    2.331060] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.7.0-rc5-01053-g43a7eab-dirty #4
[    2.331268] Hardware name: QEMU QEMU Virtual Machine, BIOS 0.0.0 02/06/2015
[    2.331585] Call trace:
<snip>
```
As I think you need to update logic in Linux TEE device registration here [1] as well to give different names to devices enumerated without supplicant dependency and with supplicant dependency.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/tee/optee/device.c#n78

Comment (muvarov) [reply core/pta/device.c:27]: I didn't setup 2 flags in one time. But yes, there is bunch of places where linux drivers needs to be changes (like you pointed to sysfs entries). I will add check for now and if there will be use case then we can rethink how to support 2 flags together.

File: core/pta/device.c [hunk 9]
```diff
@@ -21,9 +21,15 @@
 #define PTA_NAME "device.pta"
 
 static void add_ta(uint32_t flags, const TEE_UUID *uuid, uint8_t *buf,
-		   uint32_t blen, uint32_t *pos)
+		   uint32_t blen, uint32_t *pos, uint32_t rflags)
 {
-	if (flags & TA_FLAG_DEVICE_ENUM) {
+	if ((flags & TA_FLAG_DEVICE_ENUM) &&
+	    (flags & TA_FLAG_DEVICE_ENUM_SUPP)) {
+		EMSG("pta: skipping ta %pUl, broken enum flags\n", uuid);
```
Comment (jforissier) [inline core/pta/device.c:28]: I have 4 comments to make on this line ;-)
1. You could use PTA_NAME
2. s/ta/TA/
3. "broken": I prefer "inconsistent"
4. `\n` not needed

Therefore:
```C
EMSG(PTA_NAME ": skipping TA %pUl, inconsistent flags");
```

File: lib/libutee/include/user_ta_header.h [hunk 10]
```diff
@@ -23,9 +23,16 @@
 	 * (pseudo-TAs only).
 	 */
 #define TA_FLAG_CONCURRENT		(1 << 8)
-#define TA_FLAG_DEVICE_ENUM		(1 << 9) /* device enumeration */
+	/*
+	 * Device enumeration is done in two stages by normal world, first
+	 * before the tee-supplicant has started and then once more when the
+	 * tee-supplicant is started. The flags below controls if the TA should
+	 * be reported in the first, second or in both cases.
```
Comment (b49020) [inline lib/libutee/include/user_ta_header.h:30]: Here you should drop "both cases" as well.

**Comment by jforissier** [comment]
You didn't comment on my previous remark:

> I find the commit title a bit hard to understand, what does "watch for" mean in this context? I suggest "core: device pta: add flag to indicate dependency on tee-supplicant"

Anyway, this looks OK to me.
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by muvarov** [comment]
> You didn't comment on my previous remark:
> 
> > I find the commit title a bit hard to understand, what does "watch for" mean in this context? I suggest "core: device pta: add flag to indicate dependency on tee-supplicant"
> 
Missed that comment. Agree it sounds better. Will update.

> Anyway, this looks OK to me.
> `Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`
